### PR TITLE
Expose function for locating hostfxr

### PR DESCRIPTION
**Pull request by elinor-fung**
- Add nethost library
- Implement nethost_get_hostfxr_path
- Add basic test for (a native host using) nethost

### CODE REVIEW

File: src/corehost/Windows/gen-buildsys-win.bat [hunk 1]
```diff
@@ -41,8 +41,8 @@ for /f "delims=" %%a in ('powershell -NoProfile -ExecutionPolicy ByPass "& .\Win
...
+"%CMakePath%" %__sourceDir% %__SDKVersion% "-DCLI_CMAKE_RUNTIME_ID:STRING=%cm_BaseRid%" "-DCLI_CMAKE_HOST_VER:STRING=%__HostVersion%" "-DCLI_CMAKE_APPHOST_VER:STRING=%__AppHostVersion%" "-DCLI_CMAKE_COMHOST_VER:STRING=%__AppHostVersion%" "-DCLI_CMAKE_IJWHOST_VER:STRING=%__AppHostVersion%" "-DCLI_CMAKE_NETHOST_VER:STRING=%__AppHostVersion%" "-DCLI_CMAKE_HOST_FXR_VER:STRING=%__HostFxrVersion%" "-DCLI_CMAKE_HOST_POLICY_VER:STRING=%__HostPolicyVersion%" "-DCLI_CMAKE_PKG_RID:STRING=%cm_BaseRid%" "-DCLI_CMAKE_COMMIT_HASH:STRING=%__LatestCommit%" "-DCLI_CMAKE_PLATFORM_ARCH_%cm_Arch%=1" "-DCMAKE_INSTALL_PREFIX=%__CMakeBinDir%" "-DCLI_CMAKE_RESOURCE_DIR:STRING=%__ResourcesDir%" -G "Visual Studio %__VSString%" %__ExtraCmakeParams%
 endlocal
```
Comment (sdmaclea) [inline src/corehost/Windows/gen-buildsys-win.bat]: I would prefer we stop creating these reduntant defines.  Change 
```C#
"-DCLI_CMAKE_APPHOST_VER:STRING=%__AppHostVersion%" "-DCLI_CMAKE_COMHOST_VER:STRING=%__AppHostVersion%" "-DCLI_CMAKE_IJWHOST_VER:STRING=%__AppHostVersion%" "-DCLI_CMAKE_NETHOST_VER:STRING=%__AppHostVersion%"
```
to something like:
```C#
"-DCLI_CMAKE_COMMON_HOST_VER:STRING=%__AppHostVersion%"
```
/cc @vitek-karas @jkoritzinsky @AaronRobinsonMSFT Opinions?

File: src/corehost/build.sh [hunk 2]
```diff
@@ -263,9 +263,9 @@ if [ $__CrossBuild == 1 ]; then
...
 else
-    cmake "$DIR" -G "Unix Makefiles" $__cmake_defines -DCLI_CMAKE_HOST_VER:STRING=$__host_ver -DCLI_CMAKE_APPHOST_VER:STRING=$__apphost_ver -DCLI_CMAKE_HOST_FXR_VER:STRING=$__fxr_ver -DCLI_CMAKE_HOST_POLICY_VER:STRING=$__policy_ver -DCLI_CMAKE_PKG_RID:STRING=$__base_rid -DCLI_CMAKE_COMMIT_HASH:STRING=$__commit_hash -DCMAKE_INSTALL_PREFIX=$__cmake_bin_prefix
+    cmake "$DIR" -G "Unix Makefiles" $__cmake_defines -DCLI_CMAKE_HOST_VER:STRING=$__host_ver -DCLI_CMAKE_APPHOST_VER:STRING=$__apphost_ver -DCLI_CMAKE_NETHOST_VER:STRING=$__apphost_ver -DCLI_CMAKE_HOST_FXR_VER:STRING=$__fxr_ver -DCLI_CMAKE_HOST_POLICY_VER:STRING=$__policy_ver -DCLI_CMAKE_PKG_RID:STRING=$__base_rid -DCLI_CMAKE_COMMIT_HASH:STRING=$__commit_hash -DCMAKE_INSTALL_PREFIX=$__cmake_bin_prefix
 fi
```
Comment (sdmaclea) [inline src/corehost/build.sh]: If we agree to use `CLI_CMAKE_COMMON_HOST_VER`, then also here. (See above)

File: src/corehost/cli/fxr_resolver.cpp [hunk 3]
```diff
@@ -54,80 +53,90 @@ namespace
...
-            trace::info(_X("Using global installation location [%s] as runtime location."), default_install_location.c_str());
-            out_dotnet_root->assign(default_install_location);
-        }
-        else
-        {
-            trace::error(_X("A fatal error occurred, the default install location cannot be obtained."));
-            return false;
-        }
-        trace::info(_X("Using default installation location [%s] as runtime location."), default_install_location.c_str());
-        out_dotnet_root->assign(default_install_location);
```
Comment (sdmaclea) [inline src/corehost/cli/fxr_resolver.cpp]: Were these two lines intentionally rmoved/removed?

File: src/corehost/cli/fxr_resolver.cpp [hunk 4]
```diff
@@ -54,80 +53,90 @@ namespace
     }
 }
 
-bool resolve_fxr_path(const pal::string_t& root_path, pal::string_t* out_dotnet_root, pal::string_t* out_fxr_path)
+bool fxr_resolver::try_get_path(const host_mode_t mode, const pal::string_t& root_path, pal::string_t* out_dotnet_root, pal::string_t* out_fxr_path)
 {
     pal::string_t fxr_dir;
-#if FEATURE_APPHOST || FEATURE_LIBHOST
```
Comment (sdmaclea) [inline src/corehost/cli/fxr_resolver.cpp]: We go to great lengths to avoid breaking changes to the muxer.  I wonder if we should keep some #if here.  Something like:
```C++
#if !FEATURE_MUXER
```

Or maybe it would be better to leave the interface unchanged.  Keep most of the original code and interface. I see no significant functional change to this file, just refactoring for cleanup.
```C++
bool resolve_fxr_path(const pal::string_t& root_path, pal::string_t* out_dotnet_root, pal::string_t* out_fxr_path)
```

File: src/corehost/Windows/gen-buildsys-win.bat (hunk 1 above)
Comment (elinor-fung) [reply src/corehost/Windows/gen-buildsys-win.bat]: I like it (but I may be missing context/reason for why they are separate defines right now)

File: src/corehost/cli/fxr_resolver.cpp (hunk 3 above)
Comment (elinor-fung) [reply src/corehost/cli/fxr_resolver.cpp]: Yes, they are (re)doing the same thing as in the if block above.

File: src/corehost/cli/fxr_resolver.cpp (hunk 4 above)
Comment (elinor-fung) [reply src/corehost/cli/fxr_resolver.cpp]: I think I'm missing what the breaking change is here?

The functionality / logic should be the same. For nethost, we need to be able to resolve using either the muxer behaviour or the apphost/libhost behaviour, so I removed the preprocessor conditions here. But corehost.cpp is setting the mode based on the defines such that the behaviour should be the same (and com/ijwhost, which have FEATURE_LIBHOST, are always using host_mode_t::libhost)

File: src/corehost/Windows/gen-buildsys-win.bat (hunk 1 above)
Comment (jkoritzinsky) [reply src/corehost/Windows/gen-buildsys-win.bat]: I'd be fine with collapsing them. I was just following the convention of the COM host.

Comment (AaronRobinsonMSFT) [reply src/corehost/Windows/gen-buildsys-win.bat]: These are historic nonsense and have very little use for all hosts except `dotnet.exe`. Basically the `ijwhost`, `comhost`, `apphost`, and `nethost` can all have the same version since they ship and version with the SDK.

File: src/corehost/cli/fxr_resolver.cpp (hunk 4 above)
Comment (AaronRobinsonMSFT) [reply src/corehost/cli/fxr_resolver.cpp]: This isn't an exported function. There is no breaking change here.

File: src/corehost/cli/nethost/nethost.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+#define __NETHOST_H__
+
+#if defined(_WIN32)
+    #ifdef NETHOST_EXPORT
+        #define NETHOST_API __declspec(dllexport)
+    #else
+        #define NETHOST_API __declspec(dllimport)
+    #endif
+
+    #define NETHOST_CALLTYPE __stdcall
```
Comment (AaronRobinsonMSFT) [inline src/corehost/cli/nethost/nethost.h]: `__stdcall` will be an error on ARM for Windows I think.

File: src/corehost/cli/fxr_resolver.cpp (hunk 4 above)
Comment (sdmaclea) [reply src/corehost/cli/fxr_resolver.cpp]: We spent months getting approval to make changes to the muxer.  We had to justify changes to 
* Add the manifest
* Add logging to file support
* Add support for SemVer 2.0.0

We could logically reason that none of these were breaking.

This PR is introducing a non-functional change to the muxer.  I agree it shouldn't be breaking, but it does change muxer code.  It is introducing a switch and dead code which was not compiled for the muxer before.

If there was a functional reason for the code change, it is OK to keep the switch. I would still think about restoring an `#if` to eliminate the dead code for the muxer. I'll let @vitek-karas decide if I am being overly cautious.

Comment (AaronRobinsonMSFT) [reply src/corehost/cli/fxr_resolver.cpp]: >We spent months getting approval to make changes to the muxer.

Not for `comhost` which had much more impactful functional changes - the very notion of `libhost` for instance.

Comment (AaronRobinsonMSFT) [reply src/corehost/cli/fxr_resolver.cpp]: Actually manifests have OS level requirements and impact, logging to file is an operational change since it has potential perf and file system impact, and versioning impacts literally everything. I really don't see how any of these code changes are even close to the same level as those.

Comment (sdmaclea) [reply src/corehost/cli/fxr_resolver.cpp]: > Not for comhost which had much more impactful functional changes - the very notion of libhost for instance.

Only because we didn't mention they affected the muxer.  If there are changes to the muxer, we must review them carefully. @jeffschwMSFT wants us to make sure there is sufficient coverage to guarantee the muxer works properly for every shipping .NET Core version.

Comment (AaronRobinsonMSFT) [reply src/corehost/cli/fxr_resolver.cpp]: >Only because we didn't mention they affected the muxer.

That just isn't true. The impact of `comhost` on the muxer was discussed extensively.

I agree that coverage is an important consideration. I am hoping the start of the native testing here will allow us to do just that. @vitek-karas also mentioned you had started work on some kind of native testing. Is your work applicable here?

Comment (sdmaclea) [reply src/corehost/cli/fxr_resolver.cpp]: > That just isn't true

Offline

> Is your work applicable here?

The unit and E2E tests are checked in. In my opinion they were sufficient to test the SemVer 2.0.0 work.

File: src/corehost/cli/nethost/nethost.h (hunk 5 above)
Comment (AaronRobinsonMSFT) [reply src/corehost/cli/nethost/nethost.h]: Nevermind. Seems to build.

File: src/corehost/Windows/gen-buildsys-win.bat (hunk 1 above)
Comment (elinor-fung) [reply src/corehost/Windows/gen-buildsys-win.bat]: Switched to `CLI_CMAKE_COMMON_HOST_VER`

File: src/corehost/cli/test/nativehost/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# Licensed to the .NET Foundation under one or more agreements.
+# The .NET Foundation licenses this file to you under the MIT license.
+# See the LICENSE file in the project root for more information.
+
+cmake_minimum_required (VERSION 2.8.12)
+project(nativehost)
+
+set(DOTNET_PROJECT_NAME "nativehost")
+
+set(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)
```
Comment (vitek-karas) [inline src/corehost/cli/test/nativehost/CMakeLists.txt]: Should this OSX specific stuff be moved to the `testexe.cmake`? So that all our test exes share it. I admit I don't really know what this code does, but if we need it for native host, we might need it for other tests as well.

File: src/corehost/cli/test/nativehost/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+include_directories(${CMAKE_CURRENT_LIST_DIR}/../../nethost)
+
+set(SOURCES
+    ./nativehost.cpp
+)
+
+include(../testexe.cmake)
+
+set_common_libs("exe")
```
Comment (vitek-karas) [inline src/corehost/cli/test/nativehost/CMakeLists.txt]: Can we move this to `testexe.cmake` as well?

File: src/corehost/corehost.cpp [hunk 8]
```diff
@@ -154,9 +154,17 @@ int exe_start(const int argc, const pal::char_t* argv[])
     app_path.append(_X(".dll"));
 #endif
 
+    host_mode_t mode;
+#if FEATURE_APPHOST
+    mode = host_mode_t::apphost;
+#elif FEATURE_LIBHOST
```
Comment (vitek-karas) [inline src/corehost/corehost.cpp]: `libhost` should never get here - there's no `libhost` with `main` and there's also no `libhost` which "runs an application".

File: src/test/HostActivationTests/NativeHosting/Nethost.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+using Xunit;
+
+namespace Microsoft.DotNet.CoreSetup.Test.HostActivation.NativeHosting
+{
+    public class Nethost : IClassFixture<Nethost.SharedTestState>
+    {
+        private const string GetHostFxrPath = "nethost_get_hostfxr_path";
+
+        private readonly SharedTestState sharedState;
+        private readonly DotNetCli dotNetCli;
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/Nethost.cs]: move this as a property of the shared test state.

File: src/test/HostActivationTests/NativeHosting/Nethost.cs [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        [Fact]
+        public void GetHostFxrPath_NoAssemblyPath_NoFxr()
+        {
+            Command.Create(sharedState.NativeHostPath, GetHostFxrPath)
+                .CaptureStdErr()
+                .CaptureStdOut()
+                .Execute()
+                .Should().Fail();
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/Nethost.cs]: Can we validate the failure - exit code, error message.

File: src/corehost/cli/nethost/nethost.cpp [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+NETHOST_API int NETHOST_CALLTYPE nethost_get_hostfxr_path(
+    nethost_get_hostfxr_path_result_fn result,
+    const char_t * assembly_path)
+{
+    host_mode_t mode = host_mode_t::invalid;
+    pal::string_t root_path;
+    if (assembly_path == nullptr)
+    {
+        mode = host_mode_t::muxer;
```
Comment (vitek-karas) [inline src/corehost/cli/nethost/nethost.cpp]: Sorry - this is my bad.
We can't use the muxer behavior here. The muxer behavior assumes that the `dotnet.exe` is called and thus it searches next to it to find `hostfxr` - and nowhere else. That's effectively useless for the `nethost` scenarios, because the `nethost` will typically run in the cotext of some application, which will not have the `hostfxr` next to it.

We will need to add some more tweaks. I think the proper behavior should be:
* `assembly_path` is `nullptr`: basically the `apphost` behavior but without the "root_path", so don't look next to itself. But do use `DOTNET_ROOT` and do look in global locations.
* `assembly_path` specified: exactly like `apphost` behavior - this is already below.

File: src/test/HostActivationTests/NativeHosting/Nethost.cs [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        [Fact]
+        public void GetHostFxrPath_NoAssemblyPath_FxrSubdirectory()
+        {
+            string copiedHostPath = Path.Combine(Path.GetDirectoryName(sharedState.NativeHostPath), "host");
+            SharedFramework.CopyDirectory(Path.Combine(dotNetCli.BinPath, "host"), copiedHostPath);
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/Nethost.cs]: This will probably go away, but if we keep it in some form: The test pollutes the directory used by other tests - it puts the hostfxr next to the native host... and keeps it there for the other tests.

You need some kind of try/finally to delete it - or create a copy in a new directory just for this test.

File: src/test/HostActivationTests/NativeHosting/Nethost.cs [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+                string nativeHostName = RuntimeInformationExtensions.GetExeFileNameForCurrentPlatform("nativehost");
+                NativeHostPath = Path.Combine(baseDir, nativeHostName);
+
+                // Copy over native host and nethost
+                string nethostName = RuntimeInformationExtensions.GetSharedLibraryFileNameForCurrentPlatform("nethost");
+                File.Copy(Path.Combine(RepoDirectories.CorehostPackages, nethostName), Path.Combine(baseDir, nethostName));
+                File.Copy(Path.Combine(RepoDirectories.Artifacts, "corehost_test", nativeHostName), NativeHostPath);
+
+                var standaloneAppFixture = new TestProjectFixture("StandaloneApp", RepoDirectories);
+                StandaloneAppFixture = standaloneAppFixture
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/Nethost.cs]: I know this works, but we should try to avoid full blown build/publish unless necessary. It is really slow. For this test all we need is a fake `app.dll` and the `hostfxr` next to it.

On top of that, the `hostfxr` can also be a fake (empty file).

Let's try to make new tests as fast as possible - it already takes cca 15 minutes to run the tests, way too long.

File: src/corehost/cli/nethost/nethost.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/corehost/cli/nethost/nethost.h]: @vitek-karas Bah... Yes, you are right. The rest of the core-setup API is `__cdecl`. I agree that using `__stdcall` is more consistent with `coreclr.dll`, `comhost`, `ijwhost`, and the proposed `winrthost` exports. I will say this is annoying since all of `hostpolicy` and `hostfxr` are `__cdecl`. So much annoyance here. Thanks for pointing this out. I think we should keep with `__stdcall` since it aligns with all non "internal" core-setup scenarios better. I consider interacting directly with `hostfxr` and `hostpolicy` more "internal" or at the very least "advanced" as compared to COM, IJW, WinRT, and the native loading scenarios which users could experiment with.

[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/corehost/cli/nethost/nethost.h]: Agreed - and I like the consistency among the "public" hosts. `hostpolicy` is not going to be used externally (it's just too hard to use directly). `hostfxr` uses will be advanced and so some complexity is OK.

[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/corehost/cli/nethost/nethost.h]: @elinor-fung In a previous review @vitek-karas mentioned the need for documentation of these kinds of contracts. I think the fact that we are starting to coalesce around common user scenarios using `__stdcall` on x86 is an important contract that should be documented and officially defined. I have observed you reading through the documentation of core-setup, is there an existing doc that could contain these kind of details or is a new one needed?

File: src/corehost/cli/nethost/nethost.h [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        #define NETHOST_API __attribute__((__visibility__("default")))
+    #else
+        #define NETHOST_API
+    #endif
+
+    #define NETHOST_CALLTYPE
+    using char_t = char;
+#endif
+
+using nethost_get_hostfxr_path_result_fn = void(*)(const char_t * hostfxr_path);
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/corehost/cli/nethost/nethost.h]: > Using the callback makes it harder to use from some languages. It means the caller has to store the value in a global variable, which introduces issues.

I would agree with this sentiment.

File: src/corehost/cli/nethost/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    ../fxr/fx_ver.cpp
+)
+
+include(../lib.cmake)
+
+add_definitions(-DFEATURE_LIBHOST=1)
+add_definitions(-DNETHOST_EXPORT)
+
+install(TARGETS nethost DESTINATION corehost)
+install_symbols(nethost corehost)
```
Comment (jkoritzinsky) [inline src/corehost/cli/nethost/CMakeLists.txt]: We should add an `install()` command to install the public `nethost.h` header for easy distribution. I think the `PUBLIC_HEADER` target property and install command is available in all the CMake versions we use in this repo.

File: src/corehost/cli/fxr_resolver.cpp (hunk 4 above)
Comment (elinor-fung) [reply src/corehost/cli/fxr_resolver.cpp]: Since it turned out we didn't need the muxer behaviour in nethost after all, I went back to the preprocessor conditions.

File: src/corehost/cli/nethost/nethost.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (elinor-fung) [reply src/corehost/cli/nethost/nethost.h]: Switched to a buffer + size supplied by the caller.

File: src/corehost/cli/test/nativehost/CMakeLists.txt (hunk 6 above)
Comment (elinor-fung) [reply src/corehost/cli/test/nativehost/CMakeLists.txt]: This is setting the search path for the executable to load libraries at runtime. For the native host, I made it look right next to itself (the tests copy nethost next to it). Other executables might not need to add to the search path all or might want to add something different, so I think it makes sense to leave it out of `testexe.cmake` (at least for now - if we do end up with a bunch of test executables that want the same thing, we can move it)

File: src/corehost/cli/test/nativehost/CMakeLists.txt (hunk 7 above)
Comment (elinor-fung) [reply src/corehost/cli/test/nativehost/CMakeLists.txt]: Moved.

File: src/corehost/cli/nethost/nethost.cpp (hunk 11 above)
Comment (elinor-fung) [reply src/corehost/cli/nethost/nethost.cpp]: Updated to stop using the muxer behaviour and follow what you outlined above.

File: src/test/HostActivationTests/NativeHosting/Nethost.cs (hunk 12 above)
Comment (elinor-fung) [reply src/test/HostActivationTests/NativeHosting/Nethost.cs]: This went away, but I ended up making the shared test state just set everything up separately.

File: src/test/HostActivationTests/NativeHosting/Nethost.cs (hunk 13 above)
Comment (elinor-fung) [reply src/test/HostActivationTests/NativeHosting/Nethost.cs]: Switched to creating a fake `app.dll` / `hostfxr.dll` instead of using an actual project.

File: src/corehost/corehost.cpp (hunk 8 above)
Comment (elinor-fung) [reply src/corehost/corehost.cpp]: Got rid of passing the mode at all.

File: src/corehost/cli/nethost/CMakeLists.txt (hunk 15 above)
Comment (elinor-fung) [reply src/corehost/cli/nethost/CMakeLists.txt]: From what I read, it seems like `PUBLIC_HEADER` does different things on Apple. I made it put `nethost.h` right next to the .lib and .dll. we haven't dealt with the packaging/distribution side of things yet, but those three will need to go together, so I think at least putting them next to each other in bin makes sense for now.

File: src/corehost/cli/nethost/nethost.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (elinor-fung) [reply src/corehost/cli/nethost/nethost.h]: I'd say we need a new one. Maybe what gets created for #5481 would be a reasonable place (if it is a doc for all exposed APIs - I could also see doing a separate 'advanced' `hostfxr`/`hostpolicy` doc and 'public' `*host` doc)

File: src/corehost/cli/nethost/CMakeLists.txt (hunk 15 above)
Comment (jkoritzinsky) [reply src/corehost/cli/nethost/CMakeLists.txt]: Sounds good!

File: src/corehost/cli/nethost/nethost.cpp [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+#include <pal.h>
+#include <utils.h>
+
+NETHOST_API int NETHOST_CALLTYPE nethost_get_hostfxr_path(
+    char_t * result_buffer,
+    size_t buffer_size,
+    size_t * out_buffer_required_size,
+    const char_t * assembly_path)
+{
+    if (out_buffer_required_size == nullptr)
```
Comment (vitek-karas) [inline src/corehost/cli/nethost/nethost.cpp]: Call `trace::setup()` as the first thing, to make sure that tracing works correctly.

File: src/test/HostActivationTests/NativeHosting/Nethost.cs [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+            if (isValid)
+            {
+                result.Should().Pass()
+                    .And.HaveStdOutContaining($"hostfxr_path: {sharedState.HostFxrPath}".ToLower());
+            }
+            else
+            {
+                result.Should().Fail()
+                    .And.ExitWith(1)
+                    .And.HaveStdOutContaining($"{GetHostFxrPath} failed: 0x{CoreHostLibMissingFailure.ToString("x")}");
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/Nethost.cs]: You could additionally validate that the error written out is expected  - through enabling tracing.

File: src/corehost/cli/nethost/nethost.cpp (hunk 16 above)
Comment (vitek-karas) [reply src/corehost/cli/nethost/nethost.cpp]: Also - the code should not write anything to `stderr` which it does by default for errors. See IJW host which does the same thing here:
https://github.com/dotnet/core-setup/blob/7d83e1fa27bb2ef3e3623f37370023446a3d3dbe/src/corehost/cli/ijwhost/ijwthunk.cpp#L118

File: src/test/HostActivationTests/NativeHosting/Nethost.cs [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+        [Theory]
+        [InlineData(false, true)]
+        [InlineData(false, false)]
+        [InlineData(true, true)]
+        [InlineData(true, false)]
+        public void GetHostFxrPath_GlobalInstallation(bool useAssemblyPath, bool isValid)
+        {
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                // We don't have a good way of hooking into how the product looks for global installations yet.
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/Nethost.cs]: We will need something here eventually... 👍

File: src/test/HostActivationTests/NativeHosting/Nethost.cs [hunk 19]
```diff
@@ -88,22 +117,59 @@ public void GetHostFxrPath_WithAssemblyPath_AppLocalFxr()
...
             Command.Create(sharedState.NativeHostPath, $"{GetHostFxrPath} {assemblyPath}")
                 .CaptureStdErr()
                 .CaptureStdOut()
+                .EnvironmentVariable("COREHOST_TRACE", "1")
                 .Execute()
                 .Should().Pass()
                 .And.HaveStdOutContaining($"hostfxr_path: {hostFxrPath}".ToLower());
         }
 
+        private class TestRegistryKeyOverride : IDisposable
```
Comment (vitek-karas) [inline src/test/HostActivationTests/NativeHosting/Nethost.cs]: Can you please expose this to all tests and refactor the one in SDK resolution to use the same class? It does exactly the same thing.

Comment (elinor-fung) [reply src/test/HostActivationTests/NativeHosting/Nethost.cs]: Pulled out into `RegisteredInstallKeyOverride` class and switched other tests to it.

File: src/corehost/cli/nethost/nethost.h [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+//
+// Return value:
+//   0 on success, otherwise failure
+//   0x80008098 - result_buffer is too small (HostApiBufferTooSmall)
+//
+// Remarks:
+//   The full search for the hostfxr library is done on every call. To minimize the need
+//   to call this function multiple times, pass a large result_buffer (e.g. PATH_MAX).
+//
+extern "C" NETHOST_API int NETHOST_CALLTYPE nethost_get_hostfxr_path(
```
Comment (jkotalik) [inline src/corehost/cli/nethost/nethost.h]: Is it expected that native hosters carry this dll with them rather than it being app local? Ex if I have a global module, and I wanted to find hostfxr, I would need to have nethost.dll sit side by side with my global module?

Comment (elinor-fung) [reply src/corehost/cli/nethost/nethost.h]: Yes, the expectation is that a native host application will ship this library as part of the application itself.

Comment (jkotalik) [reply src/corehost/cli/nethost/nethost.h]: I think I need a bit of clarification. The expectation is that a published app will contain the nethost.dll, like:

```
published/
    nethost.dll
    myapp.dll
```

Also, is the nethost.dll opt-in or opt-out?

Comment (vitek-karas) [reply src/corehost/cli/nethost/nethost.h]: The `nethost.dll` is to be part of the native host - so in the case of ASP.NET it will be part of the ASP.NET IIS module.
* .NET Core app doesn't need it, since the `apphost` or muxer will find the `hostfxr` for it and if it needs the path to `hostfxr` at runtime it can simply ask for already loaded module.
* non .NET Core app which uses `hostfxr` to host .NET Core (this is what we call "native host") will include it in itself.

Basically the ASP.NET IIS module has the "locate hostfxr" code copied into it today, with this it would just carry it as a separate dll.

We don't have a delivery vehicle yet, but we're thinking native (C/C++) NuGet package and probably just raw `.zip` with the library and header file for all other cases (Linux/macOS, non-C/C++ languages).

Comment (jkotalik) [reply src/corehost/cli/nethost/nethost.h]: Our IIS module today installs via an msi, so I'd imagine we should probably include it SxS with the module. 

Once we get new bits from core-setup, I will start prototyping and see if all scenarios we used to support still work with `nethost.dll`.

Comment (vitek-karas) [reply src/corehost/cli/nethost/nethost.h]: @jkotalik Just to clarify for me
> include it SxS with the module

This means that you would package the `nethost` with your module and install it into the same folder on the target machine... right?

Re the compatibility - if that's not the case (that is you find a case where it works differently then your existing code), please let us know. It's using the same code as `apphost` and so if there's a problem in `nethost` it's likely there is a problem in the other "hosts" as well.

Comment (jkotalik) [reply src/corehost/cli/nethost/nethost.h]: > This means that you would package the nethost with your module and install it into the same folder on the target machine... right?

Yes I believe that is the best course of action.

> Re the compatibility - if that's not the case (that is you find a case where it works differently then your existing code), please let us know. It's using the same code as apphost and so if there's a problem in nethost it's likely there is a problem in the other "hosts" as well.

I'm less concerned about that than some of the backups we did. For example, if all else fails with finding dotnet, we search in C:\PF\dotnet as a backup. So we will see!
